test/integration/address_cache.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+
+"""Test recent address cache."""
+
+import os
+import typing as T
+import shutil
+
+from testlib import check
+from testlib.log import log
+from testlib.proc import Tinc, Script
+from testlib.test import Test
+
+
+def init(ctx: Test) -> T.Tuple[Tinc, Tinc]:
+    """Create test node."""
+    foo, bar = ctx.node(), ctx.node()
+
+    stdin = f"""
+        init {foo}
+        set Port 0
+        set Address localhost
+        set DeviceType dummy
+    """
+    foo.cmd(stdin=stdin)
+
+    return foo, bar
+
+
+def run_tests(ctx: Test) -> None:
+    """Run tests."""
+    foo, bar = init(ctx)
+
+    log.info("cache directory must exist after init")
+    check.dir_exists(foo.sub("cache"))
+
+    foo.add_script(Script.INVITATION_ACCEPTED)
+    foo.start()
+
+    log.info("invite %s to %s", bar, foo)
+    invite, _ = foo.cmd("invite", bar.name)
+    invite = invite.strip()
+
+    log.info("join %s to %s", bar, foo)
+    bar.cmd("join", invite)
+
+    log.info("invitee address must be cached after invitation is accepted")
+    foo[Script.INVITATION_ACCEPTED].wait()
+    check.file_exists(foo.sub(f"cache/{bar}"))
+    os.remove(foo.sub(f"cache/{bar}"))
+
+    log.info("configure %s", bar)
+    bar.cmd("set", "DeviceType", "dummy")
+    bar.cmd("set", "Port", "0")
+
+    log.info("connect nodes")
+    bar.add_script(foo.script_up)
+    bar.cmd("start")
+    bar[foo.script_up].wait()
+
+    log.info("cache files must exist after connection is established")
+    check.file_exists(bar.sub(f"cache/{foo}"))
+
+    log.info("stop nodes")
+    bar.cmd("stop")
+
+    log.info("remove %s cache directory", bar)
+    shutil.rmtree(bar.sub("cache"))
+
+    log.info("start node %s", bar)
+    bar.cmd("start")
+    bar[foo.script_up].wait()
+
+    log.info("make sure %s cache was recreated", bar)
+    check.file_exists(bar.sub(f"cache/{foo}"))
+
```

Comment:
It might also be nice to actually check that the address cache contains useful information, by removing stopping the nodes, removing `Address`es from the all host config files, and then seeing if they can reconnect just from the information in the cache.

Reply:
Makes sense, added.